### PR TITLE
#356 add tebako-pack workflow

### PR DESCRIPTION
**Pull request by CAMOBAP**
In development

### CODE REVIEW

**Comment by CAMOBAP** [comment]
@maxirmx maybe you can have a quick look the failure

For some reason `gflags` compilation starts once I moved `tebako` related steps to separate action (probably we may make it reusable)

- https://github.com/fontist/fontist/actions/runs/7688363054/job/20949498456 - works with configuration in workflow
- https://github.com/fontist/fontist/actions/runs/7688949709/job/20950767475 - doesn't work with 'local' `setup-tebako`

**Comment by maxirmx** [comment]
> @maxirmx maybe you can have a quick look the failure
> 
> For some reason `gflags` compilation starts once I moved `tebako` related steps to separate action (probably we may make it reusable)
> 
> * https://github.com/fontist/fontist/actions/runs/7688363054/job/20949498456 - works with configuration in workflow
> * https://github.com/fontist/fontist/actions/runs/7688949709/job/20950767475 - doesn't work with 'local' `setup-tebako`

It builds gflags in the second case as well but asks to ```Try a different combination of GFLAGS_SHARED and GFLAGS_NOTHREADS.```

Let me do it.

**Comment by maxirmx** [comment]
@CAMOBAP 

In the run that does not work you do not have "apt install" step, in the run that works you execute "apt install"
Is there any reason for it ?

**Comment by CAMOBAP** [comment]
> Is there any reason for it ?

Nope, thanks a lot you actually catched the issue

BTW how do you think in general about this approach with setup-tebako action?

It may look a Frankenstein-like at the moment but to me it looks like good approach especially it we decided to pack more gems

Any feedback appreciated

**Comment by maxirmx** [comment]
> BTW how do you think in general about this approach with setup-tebako action?
> 
> It may look a Frankenstein-like at the moment but to me it looks like good approach especially it we decided to pack more gems

Tebako has very heavy but reusble setup stage and relatively lightwiighted press stage
If we want to package more gems I would rather create docker containers (https://github.com/tamatebako/tebako/issues/67) with 
```
apt/apk/brew install
gem install tebako
tebako setup
```

and then have tebako-press action that utilizes containers 

I was delaying this step until something else then metanorma is packaged by someone ant not me.
Now when this milestone is reached I suggest I will create containers and create tebako-press action or call you to help :)
Does it sounf like a plan ?

**Comment by CAMOBAP** [comment]
@maxirmx will a such container be able to build macOS images too?

**Comment by maxirmx** [comment]
> @maxirmx will a such container be able to build macOS images too?

We will need three separate containers to build Ubuntu, MacOS and Alpine images. Tebako can onlly build Mac image on MacOs, Ubutu on Ubuntu and so on ....

**Comment by maxirmx** [comment]
Ah, we cannot have MacOS container ...

**Comment by maxirmx** [comment]
IMHO We need to package output of setup step as gem native extension. It will require some effort but will be a nice solution.

**Comment by CAMOBAP** [comment]
> IMHO We need to package output of setup step as gem native extension. It will require some effort but will be a nice solution.

Sounds like the best possible solution to me

**Comment by ronaldtse** [comment]
Agree with the move. And I agree that’s a very good solution.

File: .github/actions/setup-tebako/action.yml [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            echo "CXX=clang++" >> $GITHUB_ENV
+            echo "VCPKG_FORCE_SYSTEM_BINARIES=1" >> $GITHUB_ENV
+        else
+            echo "Unsupported operating system."
+        fi
+
+        echo "BUILD_TYPE=Release" >> $GITHUB_ENV
+        echo "TEBAKO_DIR=.archive/tebako" >> $GITHUB_ENV
+        echo "INCBIN_TAG=348e36b" >> $GITHUB_ENV
+        echo "DWARFS_WR_TAG=v0.4.0" >> $GITHUB_ENV
```
Comment (maxirmx) [inline .github/actions/setup-tebako/action.yml:55]: Please do not set ```INCBIN_TAG``` and ```DWARFS_WR_TAG```, it is a rudimant in packed_mn workflow.
```tebako``` will use some reasonable defaults

File: .github/actions/setup-tebako/action.yml [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+    - name: Install apt packages
+      if: runner.os == 'Linux'
```
Comment (maxirmx) [inline .github/actions/setup-tebako/action.yml:58]: This will work for Ubuntu, but not for Alpine

File: .github/workflows/tebako-pack.yml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    - uses: actions/checkout@v3
+      with:
+        fetch-depth: 1
+
+    - name: Setup Tebako
+      uses: ./.github/actions/setup-tebako
+      with:
+        cache: build
+        ruby_version: ${{ env.RUBY_VER }}
+        bundler_version: 2.3.22
```
Comment (maxirmx) [inline .github/workflows/tebako-pack.yml:34]: I believe it is necessary to set bundler version only for Ruby 3.0.x, 2.7.x
3.1.x, 3.2.x shall worl with defaults.  Can we try to use default here ?

**Review by maxirmx — CHANGES_REQUESTED** [review]
Asking for simplification ...
@CAMOBAP, I can play with it myself if you like.

File: .github/actions/setup-tebako/action.yml (hunk 1 above)
Comment (CAMOBAP) [reply .github/actions/setup-tebako/action.yml]: Fixed

File: .github/actions/setup-tebako/action.yml (hunk 2 above)
Comment (CAMOBAP) [reply .github/actions/setup-tebako/action.yml:56]: Fixed

File: .github/workflows/tebako-pack.yml (hunk 3 above)
Comment (CAMOBAP) [reply .github/workflows/tebako-pack.yml]: Removed

**Review by maxirmx — APPROVED** [review]
Thank you

**Comment by CAMOBAP** [comment]
To be updated after https://github.com/tamatebako/tebako/pull/124 merge

**Review by alexeymorozov — APPROVED** [review]
Looks good to me! Let's use it when finished :)

**Comment by CAMOBAP** [comment]
@maxirmx need your help with two issues in scope of this PR:

 - Ubuntu | "Could not find gems matching 'extract_ttc' valid for all resolution platforms" - https://github.com/fontist/fontist/actions/runs/7799999245/job/21276408656#step:4:6064
- macOS | "Segmentation fault: 11  ./fontist help" - https://github.com/fontist/fontist/actions/runs/7799999245/job/21276408931#step:4:12859

Maybe you can have a quick look and point me to direction to investigate

**Comment by maxirmx** [comment]
**Ubuntu issue** is related to bundler version. 
It attempts to bundle with 2.3.26 (because of  "~~> 2.3"  in fontist gemspec) and it works incorrectly.
I guess if gemspec is updated to "~> 2.4"  or bundler dependency is ommited it will work.
(if so, it is a newly introduced tebako issue)

I do not understand **MacOs issue** 
Probably I will need to debug it.

**Comment by maxirmx** [comment]
FYI  https://github.com/tamatebako/tebako/issues/126

**Comment by maxirmx** [comment]
@CAMOBAP  Gemspec setting overwrites any tebako parameters. I will try to make a fix  in tebako and ping you

**Comment by maxirmx** [comment]
@CAMOBAP 
- I fixed bundler version handling (in tebako). Now Ubuntu version works.
- MacOS package was crashing because the executable was stripped. Not sure why it causes an issue but do not strip packed_mn either
- Also I changed the ranner from macos-11 to macos-12.  Homebrew is unsupported and very slow on macos-11 and macos-11 is not supported as well. I believe it will work with macos-11 but the packaging takes an hour or so.

**Comment by CAMOBAP** [comment]
@maxirmx outstanding work!

Let me update checkout/cache steps to v4 and I think we are good to merge

**Pull request merged**